src/tqec/templates/tiled.py

```diff
@@ -0,0 +1,163 @@
+from copy import deepcopy
+from typing import Sequence
+
+import numpy
+import numpy.typing as npt
+from typing_extensions import override
+
+from tqec.exceptions import TQECException
+from tqec.position import Displacement, Position2D, Shape2D
+from tqec.templates.base import RectangularTemplate
+from tqec.templates.enums import TemplateSide
+from tqec.templates.scale import Scalable2D
+
+
+class TiledTemplate(RectangularTemplate):
+    def __init__(
+        self,
+        template_by_position: dict[Position2D, RectangularTemplate],
+        k: int = 2,
+        default_increments: Displacement | None = None,
+    ) -> None:
+        """A template representing a tiling of other templates.
+
+        The tiles are indexed by their position in the tiling.
+        """
+        super().__init__(k, default_increments)
+        if not template_by_position:
+            raise TQECException(
+                "Cannot create a tiled template with an empty template map."
+            )
+
+        scalable_shapes = {
+            template.scalable_shape for template in template_by_position.values()
+        }
+        if len(scalable_shapes) != 1:
+            raise TQECException(
+                "All templates in the TiledTemplate should have the same scalable shape."
+            )
+        self._base_scalable_shape = scalable_shapes.pop()
+
+        # Build plaquette indices map
+        indices_map: dict[Position2D, dict[int, int]] = {}
+        index_count = 1
+        indices_map = {}
+        for position, template in template_by_position.items():
+            indices_map[position] = {
+                i: i + index_count for i in range(template.expected_plaquettes_number)
+            }
+            index_count += template.expected_plaquettes_number
+        self._indices_map = indices_map
+
+        all_positions = list(template_by_position.keys())
+        min_x = min(position.x for position in all_positions)
+        max_x = max(position.x for position in all_positions)
+        min_y = min(position.y for position in all_positions)
+        max_y = max(position.y for position in all_positions)
+        # Shift the bounding box to the origin
+        self._origin_shift = Displacement(min_x, min_y)
+        self._nx = max_x - min_x + 1
+        self._ny = max_y - min_y + 1
+
+        self._template_by_position = deepcopy(template_by_position)
+        self.scale_to(k)
+
+    def get_indices_map_for_instantiation(
+        self,
+        instantiate_indices: Sequence[int] | None = None,
+    ) -> dict[Position2D, dict[int, int]]:
+        if instantiate_indices is None:
+            instantiate_indices = list(range(1, self.expected_plaquettes_number + 1))
+        index_count = 0
+        indices_map = {}
+        for position, template in self._template_by_position.items():
+            indices_map[position] = {
+                i: i + index_count
```

Comment:
Isn't
```suggestion
                i: instantiate_indices[i + index_count]
```
the correct way of doing that?

Reply:
I think it's `i: instantiate_indices[i - 1 + index_count]`.